common/persistence/dataInterfaces.go

```diff
@@ -2310,6 +2317,52 @@ func (e *WorkflowExecutionInfo) GetCurrentBranch() []byte {
 	return e.BranchToken
 }
 
+// UpdateWorkflowStateCloseStatus update the workflow state
+func (e *WorkflowExecutionInfo) UpdateWorkflowStateCloseStatus(
+	state int,
+	closeStatus int,
+) error {
+	invalidErr := &workflow.InternalServiceError{
+		Message: fmt.Sprintf(invalidStateTransitionMsg, e.State, state, closeStatus),
+	}
+
+	switch e.State {
+	case WorkflowStateVoid:
+		if state == WorkflowStateCompleted || closeStatus != WorkflowCloseStatusNone {
+			return invalidErr
+		}
+	case WorkflowStateCreated:
+		if state == WorkflowStateRunning && closeStatus != WorkflowCloseStatusNone {
+			return invalidErr
+		}
+		if state == WorkflowStateCompleted && closeStatus == WorkflowCloseStatusNone {
+			// TODO may want to further assert that close status being
+			//  WorkflowCloseStatusTerminated or
+			//  WorkflowCloseStatusTimedOut
+			return invalidErr
+		}
+
+	case WorkflowStateRunning:
+		if state == WorkflowStateCompleted && closeStatus == WorkflowCloseStatusNone {
```

Comment:
looks like this state machine check is not exhaustive. Either make it exhaustive or another option is not to do the validation in persistence at all - since it's really the upper layer which understands these states. We just happen to define them here for convenience.

Reply:
good catch